### ISSUE TITLE
The active batch rebuilds its legal mask a row at a time, and once per conditioning treatment

### ISSUE THREAD

**Issue by mmmfrieddough** [mmmfrieddough/anthro-chess#284]
## What was observed

Found while working #211, in the function that change promoted to the shared
entry point for scoring a batch. Two items sit inside `active_batch`
(`src/anthro_chess/evaluation/policy.py`) and its callers, neither of which
#211 touched.

**The legal mask is built one position at a time.** `active_batch` allocates a
tensor and does a fancy-index write per enabled row:

```python
legal_mask = torch.zeros_like(active_logits, dtype=torch.bool)
for offset, legal_actions in enumerate(legal_rows):
    legal_mask[offset, torch.tensor(legal_actions, dtype=torch.long)] = True
```

Measured on this host at the evaluation defaults (batch size 8, bucket width
32, so 256 active rows per batch): **2.44 ms per batch**, against **0.59 ms**
for one flattened `index_put_` built from `repeat_interleave` — 9.5 us against
2.3 us per position. `anthro eval run` makes roughly seven passes over the
view, so at a 400-game view that is about **1.9 s per checkpoint** of pure
host work, on the command whose GPU sits under 1% busy.

**`_ScoringSession.trajectory` builds three of them per batch where only the
logits differ.** `_condition` replaces `inputs.target_rating` and nothing else,
so `action_loss_mask`, `action_targets`, `ply_indices`, `game_ids`, and
`legal_action_ids` are the same tensor objects across all three treatments —
and `legal_policy_log_probabilities` reads only `.logits` and `.legal_mask`.
The mask build, the validation, and the enabled-row scan therefore happen three
times for one answer, about **5.4 ms per batch wasted**, plus device reads.
`_batch_keys` in `checkpoint.py` then takes a fourth device read for the game
ids and ply indices those values already carry in the same order.

## Why

Both are the same shape as #211: per-position host work in front of a tensor
operation, where the accelerator waits.

## What it buys

Zero change to any reading. The mask is the same mask; the two shared fields
are the same fields.

## What would settle it

Build the mask with one `index_put_` over precomputed row and column index
vectors. Separately, let `trajectory` build the active batch once and
`dataclasses.replace(base, logits=...)` for the other two treatments, which
also retires `_batch_keys`.

**One trap for whoever takes the second half.** `ActiveBatch.ratings` is
derived from `batch.inputs.target_rating`, which is exactly what `_condition`
replaces. A shared base would carry the true-conditioning ratings under a
corrupted treatment's name. `legal_policy_log_probabilities` never reads that
field, so it is latent rather than live, but sharing the value without either
recomputing `ratings` or making the aliasing explicit leaves a field that
lies.

## Measurement context

One Linux host, 2x RTX 4090, 32 cores, CUDA. Line-level timings taken on CPU
with tensors sized to the evaluation defaults; per-checkpoint figures derived
from those and the pass count, so treat them as estimates rather than as an
end-to-end reading.

## References

- #211 — the same class of finding in the same command, which left these two

**Comment by mmmfrieddough** [comment]
Claimed for implementation by Claude Code.

Session: `c7a2b318-45da-48a5-8c08-e48ed1427cbe`

**Comment by mmmfrieddough** [comment]
Implementation is up as #288 (ready for review). Every recorded value is unchanged — `anthro eval run` over a 400-game view at two checkpoints matched leaf for leaf across 8,194 floats — so both halves landed as stated. Three things worth having on the issue itself.

**The batch shape in the measurement context is wrong, and it understates the finding.** The body derives 256 active rows per batch from batch size 8 x bucket width 32. Real batches over `lichess-blitz-2017-04-test` carry **610 enabled rows at the median** and up to 807, because the bucket width rounds a ~60-ply game up past 32. The mask build measures 1.63 -> 0.40 ms at 256 rows and 3.31 -> 0.84 ms at 512, so the per-batch saving is roughly double the estimate, and the millisecond figures in the body are for a shape that does not occur.

**The `ratings` trap is real and is closed by recomputing.** `ActiveBatch.rescored` re-reads the conditioned rating rather than carrying it, at one small stacked device read per anchor — 13% of what the shared alignment saves. The alternative the body suggests, `dataclasses.replace(base, logits=...)` at the call site, leaves the field lying and puts the aliasing rule in the caller rather than in the type that has to honor it.

**One thing the body did not name, and it is larger than either half of this issue.** `legal_policy_log_probabilities` builds one index tensor per position and `trajectory` calls it three times per batch: **4.18 ms per call** against 0.288 ms for a gather over a prebuilt index. The index is already built and discarded by the mask build this issue vectorized. Filed as #287, blocked by #288 because the carrying mechanism is `rescored`.

Measured cost of what did land, 400-game view, best/median of five interleaved rounds: primary pass 1.10/1.24 s -> 1.01/1.11 s, trajectory pass 3.16/3.41 s -> 2.46/2.48 s.
